lib/spear.ex

```diff
@@ -594,15 +591,26 @@ defmodule Spear do
         opts
       )
 
-    case GenServer.call(conn, {:request, request}, Keyword.fetch!(opts, :timeout)) do
-      {:ok, response} ->
-        Spear.Writing.decode_append_response(
-          response[:data] || <<>>,
-          opts[:raw?]
-        )
-
-      error ->
-        error
+    with {:ok, %{status: 200, headers: headers} = response} <-
+           GenServer.call(conn, {:request, request}, Keyword.fetch!(opts, :timeout)),
+         {{"grpc-status", "0"}, _} <- {List.keyfind(headers, "grpc-status", 0), response} do
+      Spear.Writing.decode_append_response(
+        response[:data] || <<>>,
+        opts[:raw?]
+      )
+    else
+      {:error, reason} ->
+        {:error, reason}
+
+      # TODO this badly needs a refactor
+      {{"grpc-status", other_status}, response} ->
+        {:error,
+         {:grpc_failure,
+          code: other_status && String.to_integer(other_status),
+          message:
+            Enum.find_value(response.headers, fn {header, value} ->
+              header == "grpc-message" && value
+            end)}}
```

Comment:
keep in mind spear is still super rough so I think it's acceptable (for now) to have a crappy bit of control flow like this sitting here

eventually this may be refactored out to `Spear.Grpc` or something to close over the details of the well-known headers we're reading